integration-tests/src/test_loop/tests/epoch_sync.rs

```diff
@@ -0,0 +1,188 @@
+
+    // Properly shut down the previous TestLoopEnv.
+    // We must preserve the tempdir, since state dumps are stored there,
+    // and are necessary for state sync to work on the new node.
+    let tempdir = TestLoopEnv { test_loop, datas: node_datas, tempdir }
+        .shutdown_and_drain_remaining_events(Duration::seconds(5));
+
+    tracing::info!("Starting new TestLoopEnv with new node");
+
+    let clients = accounts.iter().take(NUM_CLIENTS + 1).cloned().collect_vec();
+
+    let TestLoopEnv { mut test_loop, datas: node_datas, tempdir } = TestLoopBuilder::new()
+        .genesis(genesis.clone())
+        .clients(clients)
+        .stores_override(stores)
+        .test_loop_data_dir(tempdir)
+        .config_modifier(|config, _| {
+            // Enable epoch sync, and make the horizon small enough to trigger it.
+            config.epoch_sync.enabled = true;
+            config.epoch_sync.epoch_sync_horizon = 30;
+            config.epoch_sync.epoch_sync_accept_proof_max_horizon = 20;
+            // Make header sync horizon small enough to trigger it.
+            config.block_header_fetch_horizon = 8;
+            // Make block sync horizon small enough to trigger it.
+            config.block_fetch_horizon = 3;
+        })
+        .skip_warmup()
+        .build();
+
+    // Note: TestLoopEnv does not currently propagate the network info to other peers. This is because
+    // the networking layer is completely mocked out. So in order to allow the new node to sync, we
+    // need to manually propagate the network info to the new node. In this case we'll tell the new
+    // node that node 0 is available to sync from.
+    let chain0 = &test_loop.data.get(&node_datas[0].client_sender.actor_handle()).client.chain;
+    let peer_info = HighestHeightPeerInfo {
+        archival: false,
+        genesis_id: GenesisId { chain_id: genesis.config.chain_id, hash: *chain0.genesis().hash() },
+        highest_block_hash: chain0.head().unwrap().last_block_hash,
+        highest_block_height: chain0.head().unwrap().height,
+        tracked_shards: vec![],
+        peer_info: PeerInfo {
+            account_id: Some(accounts[0].clone()),
+            addr: None,
+            id: node_datas[0].peer_id.clone(),
+        },
+    };
+    node_datas[NUM_CLIENTS].client_sender.send(SetNetworkInfo(NetworkInfo {
+        connected_peers: Vec::new(),
+        highest_height_peers: vec![peer_info], // only this field matters.
+        known_producers: vec![],
+        num_connected_peers: 0,
+        peer_max_count: 0,
+        received_bytes_per_sec: 0,
+        sent_bytes_per_sec: 0,
+        tier1_accounts_data: Vec::new(),
+        tier1_accounts_keys: Vec::new(),
+        tier1_connections: Vec::new(),
+    }));
+
+    // Check that the new node will reach a high height as well.
+    let new_node = node_datas.last().unwrap().client_sender.actor_handle();
+    test_loop.set_every_event_callback(move |test_loop_data| {
```

Comment:
If we are already checking the new node catches upto old nodes in the `run_until` below, why do we need this block?

Reply:
It's for debugging. When things don't work in a huge test, it's very useful to know in the visualizer where the node is, when looking at a specific event.